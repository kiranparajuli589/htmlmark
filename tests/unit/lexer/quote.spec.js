import Lexer from "../../../lib/lexer/index.js"


describe("quote", () => {
	it("should chill with big quote head", () => {
		const lines = [
			"> > > three hit",
			"> > two hit",
			">",
			">",
			"> > two",
			"> > # h-two",
			"> > > three",
			"> > > > four",
			"some"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should lex multiline quote with the circular depth", () => {
		const lines = [
			"> one",
			"> > two",
			"> > > three",
			"> > > > four",
			"> > >",
			">>> r_three",
			"> >",
			"> > r_two",
			">",
			"> r_one"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it.each([
		{ quote: "> quote 1", expectedDepth: 1 },
		{ quote: "> > quote 2", expectedDepth: 2 },
		{ quote: "> > > quote 3", expectedDepth: 3 },
		{ quote: "> > > > quote 4", expectedDepth: 4 },
		{ quote: "> > > > > quote 5", expectedDepth: 5 }
	])("should detect the quote depth", ({ quote, expectedDepth }) => {
		const lexer = new Lexer([quote])
		const tokenizedContent = lexer.run()
		expect(tokenizedContent[0].depth).toEqual(expectedDepth)
	})
	it.each([
		"> quote **one** with *two*",
		"  > quote `four` with ~~five~~",
		"> quote [link-title](link-url) with *two*",
		"> > quote **one** with *two*",
		"",
		" > > > quote `four` with ~~five~~",
		"> > > > quote [link-title](link-url) with *two*"
	])("should deep tokenize quote '%s'", (line) => {
		const lexer = new Lexer([line])
		const tokenizedContent = lexer.run()
		expect(tokenizedContent).toMatchSnapshot()
	})
	it("should allow laziness for plain paragraph", () => {
		const lines = [
			"> one quote",
			"> two quote",
			"three quote"
		]
		const lexer = new Lexer(lines)
		const tokenizedContent = lexer.run()
		expect(tokenizedContent).toMatchSnapshot()
	})
	it("should split with new lines in between", () => {
		const lines = [
			"> one quote",
			"",
			"> two quote",
			"three quote"
		]
		const lexer = new Lexer(lines)
		const tokenizedContent = lexer.run()
		expect(tokenizedContent).toMatchSnapshot()
	})
	it("should split with heading in between", () => {
		const lines = [
			"> one quote",
			"> two quote",
			"# three quote"
		]
		const lexer = new Lexer(lines)
		const tokenizedContent = lexer.run()
		expect(tokenizedContent).toMatchSnapshot()
	})
	it("should leave separators", () => {
		const lines = [
			"> one",
			"> # two",
			">",
			">",
			">",
			"> three"
		]
		const lexer = new Lexer(lines)
		const tokenizedContent = lexer.run()
		expect(tokenizedContent).toMatchSnapshot()
	})
	it.each([
		{ lines: [
			"> one",
			"> # two",
			"> three"
		] },
		{ lines: [
			"> # one",
			"> two",
			"> three"
		] }
	])("should not be lazy with heading", ({ lines }) => {
		const lexer = new Lexer(lines)
		const tokenizedContent = lexer.run()
		expect(tokenizedContent).toMatchSnapshot()
	})
	describe("list inside", () => {
		it("should identify list inside the quote", () => {
			const lines = [
				"> - one",
				"> - two"
			]
			const lexer = new Lexer(lines)
			const tokenizedContent = lexer.run()
			expect(tokenizedContent).toMatchSnapshot()
		})
		it("should identify list inside with single separator in between", () => {
			const lines = [
				"> - one",
				">",
				"> - two"
			]
			const lexer = new Lexer(lines)
			const tokenizedContent = lexer.run()
			expect(tokenizedContent).toMatchSnapshot()
		})
		it("should identify list inside with multiple separator in between", () => {
			const lines = [
				"> - one",
				">",
				">",
				"> - two"
			]
			const lexer = new Lexer(lines)
			const tokenizedContent = lexer.run()
			expect(tokenizedContent).toMatchSnapshot()
		})
		it("should include a braced codeblock inside", () => {
			const lines = [
				"> normal text",
				"> ```",
				"> code goes here",
				"> ```",
				"> normal text again"
			]
			const lexer = new Lexer(lines)
			const tokenizedContent = lexer.run()
			expect(tokenizedContent).toMatchSnapshot()
		})
	})
})
