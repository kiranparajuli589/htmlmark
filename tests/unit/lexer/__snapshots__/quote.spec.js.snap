// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quote list inside should identify list inside the quote 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> - one
> - two",
    "tokens": Array [
      Object {
        "indent": 0,
        "items": Array [
          Object {
            "checked": null,
            "count": null,
            "raw": " - one",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": "one",
                "type": "paragraph",
                "value": "one",
              },
            ],
            "type": "list-item",
          },
          Object {
            "checked": null,
            "count": null,
            "raw": " - two",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": "two",
                "type": "paragraph",
                "value": "two",
              },
            ],
            "type": "list-item",
          },
        ],
        "meta": Object {
          "check": false,
          "identifier": "-",
          "ordered": false,
        },
        "raw": " - one
 - two",
        "type": "list",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote list inside should identify list inside with multiple separator in between 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> - one
>
>
> - two",
    "tokens": Array [
      Object {
        "indent": 0,
        "items": Array [
          Object {
            "checked": null,
            "count": null,
            "raw": " - one",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": "one",
                "type": "paragraph",
                "value": "one",
              },
            ],
            "type": "list-item",
          },
          Object {
            "checked": null,
            "count": null,
            "raw": " - two",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": "two",
                "type": "paragraph",
                "value": "two",
              },
            ],
            "type": "list-item",
          },
        ],
        "meta": Object {
          "check": false,
          "identifier": "-",
          "ordered": false,
        },
        "raw": " - one
 - two",
        "type": "list",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote list inside should identify list inside with single separator in between 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> - one
>
> - two",
    "tokens": Array [
      Object {
        "indent": 0,
        "items": Array [
          Object {
            "checked": null,
            "count": null,
            "raw": " - one",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": "one",
                "type": "paragraph",
                "value": "one",
              },
            ],
            "type": "list-item",
          },
          Object {
            "checked": null,
            "count": null,
            "raw": " - two",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": "two",
                "type": "paragraph",
                "value": "two",
              },
            ],
            "type": "list-item",
          },
        ],
        "meta": Object {
          "check": false,
          "identifier": "-",
          "ordered": false,
        },
        "raw": " - one
 - two",
        "type": "list",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote list inside should include a braced codeblock inside 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> normal text
> \`\`\`
> code goes here
> \`\`\`
> normal text again",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " normal text",
        "type": "paragraph",
        "value": "normal text",
      },
      Object {
        "indent": 0,
        "language": null,
        "raw": " \`\`\`
 code goes here
 \`\`\`",
        "type": "code-block",
        "value": "code goes here",
      },
      Object {
        "indent": 0,
        "raw": " normal text again",
        "type": "paragraph",
        "value": "normal text again",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should allow laziness for plain paragraph 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> one quote
> two quote
three quote",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " one quote
 two quote
three quote",
        "type": "paragraph",
        "value": "one quote two quote three quote",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should chill with big quote head 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> > > three hit
> > two hit
>
>
> > two
> > # h-two
> > > three
> > > > four
some",
    "tokens": Array [
      Object {
        "depth": 1,
        "indent": 0,
        "raw": " > > three hit
 > two hit",
        "tokens": Array [
          Object {
            "depth": 1,
            "indent": 0,
            "raw": " > three hit
 two hit",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": " three hit
 two hit",
                "type": "paragraph",
                "value": "three hit two hit",
              },
            ],
            "type": "quote",
          },
        ],
        "type": "quote",
      },
      Object {
        "type": "new-line",
      },
      Object {
        "depth": 1,
        "indent": 0,
        "raw": " > two
 > # h-two
 > > three
 > > > four
some",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": " two",
            "type": "paragraph",
            "value": "two",
          },
          Object {
            "indent": 0,
            "level": 1,
            "raw": " # h-two",
            "type": "heading",
            "value": "h-two",
          },
          Object {
            "depth": 1,
            "indent": 0,
            "raw": " > three
 > > four
some",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": " three",
                "type": "paragraph",
                "value": "three",
              },
              Object {
                "depth": 1,
                "indent": 0,
                "raw": " > four
some",
                "tokens": Array [
                  Object {
                    "indent": 0,
                    "raw": " four
some",
                    "type": "paragraph",
                    "value": "four some",
                  },
                ],
                "type": "quote",
              },
            ],
            "type": "quote",
          },
        ],
        "type": "quote",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should deep tokenize quote '  > quote \`four\` with ~~five~~' 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "  > quote \`four\` with ~~five~~",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " quote \`four\` with ~~five~~",
        "type": "paragraph",
        "value": "quote \`four\` with ~~five~~",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should deep tokenize quote ' > > > quote \`four\` with ~~five~~' 1`] = `
Array [
  Object {
    "depth": 3,
    "indent": 0,
    "raw": " > > > quote \`four\` with ~~five~~",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " quote \`four\` with ~~five~~",
        "type": "paragraph",
        "value": "quote \`four\` with ~~five~~",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should deep tokenize quote '' 1`] = `Array []`;

exports[`quote should deep tokenize quote '> > > > quote [link-title](link-url) with *two*' 1`] = `
Array [
  Object {
    "depth": 4,
    "indent": 0,
    "raw": "> > > > quote [link-title](link-url) with *two*",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " quote [link-title](link-url) with *two*",
        "type": "paragraph",
        "value": "quote [link-title](link-url) with *two*",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should deep tokenize quote '> > quote **one** with *two*' 1`] = `
Array [
  Object {
    "depth": 2,
    "indent": 0,
    "raw": "> > quote **one** with *two*",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " quote **one** with *two*",
        "type": "paragraph",
        "value": "quote **one** with *two*",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should deep tokenize quote '> quote **one** with *two*' 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> quote **one** with *two*",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " quote **one** with *two*",
        "type": "paragraph",
        "value": "quote **one** with *two*",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should deep tokenize quote '> quote [link-title](link-url) with *two*' 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> quote [link-title](link-url) with *two*",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " quote [link-title](link-url) with *two*",
        "type": "paragraph",
        "value": "quote [link-title](link-url) with *two*",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should leave separators 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> one
> # two
>
>
>
> three",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " one",
        "type": "paragraph",
        "value": "one",
      },
      Object {
        "indent": 0,
        "level": 1,
        "raw": " # two",
        "type": "heading",
        "value": "two",
      },
      Object {
        "type": "new-line",
      },
      Object {
        "indent": 0,
        "raw": " three",
        "type": "paragraph",
        "value": "three",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should lex multiline quote with the circular depth 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> one
> > two
> > > three
> > > > four
> > >
>>> r_three
> >
> > r_two
>
> r_one",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " one",
        "type": "paragraph",
        "value": "one",
      },
      Object {
        "depth": 1,
        "indent": 0,
        "raw": " > two
 > > three
 > > > four
 > >
>> r_three
 >
 > r_two",
        "tokens": Array [
          Object {
            "indent": 0,
            "raw": " two",
            "type": "paragraph",
            "value": "two",
          },
          Object {
            "depth": 1,
            "indent": 0,
            "raw": " > three
 > > four
 >
> r_three",
            "tokens": Array [
              Object {
                "indent": 0,
                "raw": " three",
                "type": "paragraph",
                "value": "three",
              },
              Object {
                "depth": 1,
                "indent": 0,
                "raw": " > four",
                "tokens": Array [
                  Object {
                    "indent": 0,
                    "raw": " four",
                    "type": "paragraph",
                    "value": "four",
                  },
                ],
                "type": "quote",
              },
              Object {
                "type": "new-line",
              },
              Object {
                "indent": 0,
                "raw": " r_three",
                "type": "paragraph",
                "value": "r_three",
              },
            ],
            "type": "quote",
          },
          Object {
            "type": "new-line",
          },
          Object {
            "indent": 0,
            "raw": " r_two",
            "type": "paragraph",
            "value": "r_two",
          },
        ],
        "type": "quote",
      },
      Object {
        "type": "new-line",
      },
      Object {
        "indent": 0,
        "raw": " r_one",
        "type": "paragraph",
        "value": "r_one",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should not be lazy with heading 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> one
> # two
> three",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " one",
        "type": "paragraph",
        "value": "one",
      },
      Object {
        "indent": 0,
        "level": 1,
        "raw": " # two",
        "type": "heading",
        "value": "two",
      },
      Object {
        "indent": 0,
        "raw": " three",
        "type": "paragraph",
        "value": "three",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should not be lazy with heading 2`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> # one
> two
> three",
    "tokens": Array [
      Object {
        "indent": 0,
        "level": 1,
        "raw": " # one",
        "type": "heading",
        "value": "one",
      },
      Object {
        "indent": 0,
        "raw": " two
 three",
        "type": "paragraph",
        "value": "two three",
      },
    ],
    "type": "quote",
  },
]
`;

exports[`quote should split with heading in between 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> one quote
> two quote",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " one quote
 two quote",
        "type": "paragraph",
        "value": "one quote two quote",
      },
    ],
    "type": "quote",
  },
  Object {
    "indent": 0,
    "level": 1,
    "raw": "# three quote",
    "type": "heading",
    "value": "three quote",
  },
]
`;

exports[`quote should split with new lines in between 1`] = `
Array [
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> one quote",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " one quote",
        "type": "paragraph",
        "value": "one quote",
      },
    ],
    "type": "quote",
  },
  Object {
    "type": "new-line",
  },
  Object {
    "depth": 1,
    "indent": 0,
    "raw": "> two quote
three quote",
    "tokens": Array [
      Object {
        "indent": 0,
        "raw": " two quote
three quote",
        "type": "paragraph",
        "value": "two quote three quote",
      },
    ],
    "type": "quote",
  },
]
`;
